### PR TITLE
CORDA-699 Add injection or modification of memory network messages

### PR DESCRIPTION
**Pull request by cburlinchon**

### CODE REVIEW

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/MockNode.kt [hunk 1]
```diff
@@ -446,4 +453,57 @@ fun network(nodesCount: Int, action: MockNetwork.(nodes: List<StartedNode<MockNe
         val nodes = (1..nodesCount).map { _ -> it.createPartyNode() }
         action(it, nodes, notary)
     }
+}
+
+/**
+ * Extend this class in order to intercept and modify messages from the inmemory network
+ */
+open class MessagingServiceSpy : MessagingService {
```
Comment (al-r3) [inline testing/node-driver/src/main/kotlin/net/corda/testing/node/MockNode.kt]: You should be able to make this take a MessagingService in the constructor, use delegation, and then provide your overrides.

That way this class doesn't depend on the signature of all members of MessagingService
￼

Comment (cburlinchon) [reply testing/node-driver/src/main/kotlin/net/corda/testing/node/MockNode.kt]: done

File: docs/source/flow-testing.rst [hunk 2]
```diff
@@ -63,4 +63,17 @@ transaction as shown here.
...
 With regards to initiated flows (see :doc:`flow-state-machines` for information on initiated and initiating flows), the
 full node automatically registers them by scanning the CorDapp jars. In a unit test environment this is not possible so
-``MockNode`` has the ``registerInitiatedFlow`` method to manually register an initiated flow.
\ No newline at end of file
+``MockNode`` has the ``registerInitiatedFlow`` method to manually register an initiated flow.
+
+MockNetwork message manipulation
+--------------------------------
+The MockNetwork has the ability to manipulate message streams. You can use this to test your flows behaviour on corrupted
+or malicious data received.
```
Comment (mnesbit) [inline docs/source/flow-testing.rst]: Comma before or.

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/MockNode.kt [hunk 3]
```diff
@@ -446,4 +452,16 @@ fun network(nodesCount: Int, action: MockNetwork.(nodes: List<StartedNode<MockNe
         val nodes = (1..nodesCount).map { _ -> it.createPartyNode() }
         action(it, nodes, notary)
     }
+}
+
+/**
+ * Extend this class in order to intercept and modify messages from the inmemory network
```
Comment (mnesbit) [inline testing/node-driver/src/main/kotlin/net/corda/testing/node/MockNode.kt]: Extend this class in order to intercept and modify messages passing through the  [MessagingService] when using the [InMemoryNetwork].

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/MockNode.kt [hunk 4]
```diff
@@ -446,4 +452,16 @@ fun network(nodesCount: Int, action: MockNetwork.(nodes: List<StartedNode<MockNe
...
+ */
+open class MessagingServiceSpy(val messagingService: MessagingService) : MessagingService by messagingService
+
+/**
+ * Intercept or modify messages sent from the inmemory network
```
Comment (mnesbit) [inline testing/node-driver/src/main/kotlin/net/corda/testing/node/MockNode.kt]: Attach a [MessagingServiceSpy] to the [MockNode] allowing interception and modification of messages.

**Pull request merged**